### ISSUE TITLE
Remove Crashlytics?

### ISSUE THREAD

**Issue by IzzySoft** [yevhenii-kanivets/Money-Tracker#181]
Personal finances and tracker libraries don't go well together. Would you consider removing Crashlytics, for enhanced privacy? Thanks in advance!

**Comment by yevhenii-kanivets** [comment]
@IzzySoft thanks for getting in touch! We are considering user privacy first, this is why there is no backend storage for user's data (only personal Dropbox if user wants make backups).

In contrary, I don't think that removing Crashlytics will make any difference in terms of privacy because all crash data is anyway reported to Google Play Developer Console, there is no way to disable it. 

Crashlytics just allows us to get that data faster and in more consistent way, so we are able to faster fix any issues. Honestly, I don't think that removing it will bring any value to our users. Wdyt?

**Comment by IzzySoft** [comment]
Thanks for answering, @yevhenii-kanivets – but why would all crash reports go to the developer console "anyway" if they are not there?

As for my reasoning: with sensitive data involved, one shouldn't use proprietary tools to send data to any third party. You'd have to trust a company that lives on the data of users not to include things you didn't even tell them. And especially "us tin foils" don't like that (Google) connection (on my Android devices, there aren't even Google Services or any other apps from the "Google Apps" package present for that very reason). F-Droid would plainly reject packaging any apps coming with such libraries, for good reasons.

I'm aware that the majority "out there" is of the opinion they've got "nothing to hide" and thus don't care (but ask them to hand over all their credentials (or just their unlocked device) to prove their "nothing to hide" and you will discover otherwise). Still, there are people who **do** care. They'd prefer a version without those "known problematic candidates" – especially ones shipping data to "known privacy offenders / big data collectors". We greatly appreciate our apps coming without them, and don't consider apps that include such.

There are several options for that. If you don't want to drop those proprietary components altogether, you could use a special build flavor (or build variant) to create a "free APK" to be shared here (or to be build by [F-Droid](https://f-droid.org/) even – I'd then volunteer to help getting it listed there if you want); ship that to us "tin foils" (via F-Droid or attaching it here to `releases/` – in the latter case, I could take that APK into [my repo](https://apt.izzysoft.de/fdroid) for easy install/update on client devices), while keeping the "standard variant" for Playstore; both sides would be happy that way. Or you could replace proprietary analytics by privacy-friendly open-source ones, like [ACRA](https://github.com/ACRA/acra), which even uses opt-in by default. Or mix both approaches (Crashlytics for Play Store, ACRA for the "tin foils build").

Going that path you could even underline your "privacy first" approach visibly. Putting it thick: "we replaced … for enhanced privacy …" – which would convincingly show that other than just putting "we take your privacy, seriously" statements, you really mean it :wink:

Again, thanks for taking your time and considering! After all, it's a suggestion – I cannot expect you *have* to take it, but surely would very much appreciate if you did, whichever of the variants you chose.

**Comment by yevhenii-kanivets** [comment]
@IzzySoft I've got you point, but I'm not ready to spend my time dealing with your request, sorry :)

This is an open source solution, so you can at any moment fork it and assemble the build, which will correspond to your high privacy standards. You are free to publish it anywhere, you think it makes sense.

In the meantime I'll continue to work on the features for simple Google-friendly users. But you still can fork this repo at any time and make it high privacy compatible.

**Comment by IzzySoft** [comment]
Got your point, too – though while I personally can fork I would get no further (I'm no Android dev).

I take it you would not be opposed if someone provided a PR including the build variant? It might need minimal changes only (not being a dev, I cannot say that with confidence but just by guessing), say some lines in the `build.gradle` and one or two conditionals elsewhere. If that's correct: Mind keeping this issue open with a "help wanted" label to attract potential collaborators?

**Comment by yevhenii-kanivets** [comment]
@IzzySoft yes, we can do that 😉

**Comment by IzzySoft** [comment]
Thanks a lot! I'll try to [reach out on Mastodon](https://mastodon.technology/@IzzyOnDroid/102689436826991951) if there's an Android dev who would like to help out, pointing to this issue. Even if no one turns up straight away, that'd be at least drawing a little attention to your project – and some "boosts" might bring someone in finally. 🤞

**Comment by IzzySoft** [comment]
See: not an hour went by, and we already have almost 20 boosts. And one friendly dev who shares my opinion of "This would probably be an easy thing to do", and considers taking a look 👍

**Comment by Yky** [comment]
@IzzySoft Here you go. I created a PR. Checked with Classyshark, Crashlytics is gone.

**Comment by IzzySoft** [comment]
Phantastic, @Yky – thanks a lot!

@yevhenii-kanivets that was fast, wasn't it? Mastodon even surprised me here (and especially Yky did). When you found time to take a look and (hopefully) merge it, please let me know if you're fine with pushing that to F-Droid (and let F-Droid build the "tracker-free APK") – or if you'd prefer to build the APK yourself and have it in my repo. Pros/Cons: Faster updates via my repo – more prestige and reputation in the official one (mine is what Debian would consider a mix of non-free and testing).

**Comment by Yky** [comment]
I'm not sure @yevhenii-kanivets will want to marge this as it would remove Crashlytics from his app. 

However, @IzzySoft we could build a version from my fork and put that into F-Droid, as @yevhenii-kanivets recommends in his comment above.

**Comment by IzzySoft** [comment]
I didn't check your PR – didn't you establish a separat build flavor for that? If you "overwrote" the default flavor, it might well be that Yevhenii doesn't want to merge that.

Sure you could create a fork. But would you keep that up-to-date in a timely manner? The big advantage of a build flavor would be that once it's done, it's done and new versions would automatically profit from it.

**Comment by fynngodau** [comment]
PR #183 only removes Crashlytics from this application. Of course it would be possible to build an fdroid version from @Yky's fork, but maintaining two different source trees is tiresome and error-prone. A much cleaner way – what @IzzySoft is talking about – is to use [build flavors][build flavors], allowing to build different versions of an app from the same source tree. As a quick research showed, it is also possible to [only include a dependency in some build flavors][so], which should be done here.

[build flavors]: https://developer.android.com/studio/build/build-variants
[so]: https://stackoverflow.com/a/51626616

**Comment by yevhenii-kanivets** [comment]
@Yky @IzzySoft @fynngodau thanks for your commitment, but you are right - removing Crashlytics isn't smth I will to merge, sorry :/ Build flavour would be OK, but will take more time and efforts to do.

**Comment by IzzySoft** [comment]
We're working on it, @yevhenii-kanivets (or rather @fynngodau is giving it a try). No promises yet, I don't want to put pressure up on anyone :wink:

**Comment by fynngodau** [comment]
I have opened a pull request which will add different build flavors, see #184.

**Comment by fynngodau** [comment]
@IzzySoft In the `build.gradle` file, I have noticed that `com.dropbox.core:dropbox-core-sdk:3.0.5` is `compile`d for this application. I'm assuming this is also a proprietary library and disqualifies the app from inclusion in the main fdroid repo?

**Comment by yevhenii-kanivets** [comment]
@fynngodau thanks, I'll take a look on your PR tomorrow. FYI, Dropbox is used for one-single feature, which can be disabled by hiding a menu item in Navigation Drawer.

**Comment by fynngodau** [comment]
@yevhenii-kanivets Even if it is not used, if I'm right, no proprietary library must be included in applications that want to qualify for inclusion in the main F-Droid repository (please confirm @IzzySoft). But a proxy class could be written the same way that Google Analytics is dealt with in the PR.

**Comment by IzzySoft** [comment]
If built from source, Dropbox should be OK AFAIR as it uses a free license – as long as that license is compatible with the one of the main project (by the way: what license does Money-Tracker use?). We have some such projects at F-Droid which use Dropbox, so we can compare with how it's done there. Maybe it's available via a whitelisted Maven repo, or the code is integrated via a git submodule, or as `srclib` at F-Droid. Hard to look from here with just a browser. Just found [Money Manager Ex](https://github.com/moneymanagerex/android-money-manager-ex) is one such app according to its description, as is [MobilePauker](https://github.com/dfrits/MobilePauker2) (both using GPL3). The latter is a good match: uses it the same way as Money-Tracker (`com.dropbox.core` from JCenter).

TL;DR: Seems all fine concerning Dropbox, no changes needed except declaring a license for Money-Tracker itself. As said, GPL3 would be fine.

**Comment by fynngodau** [comment]
Oops, I'll undo the second commit then.

**Comment by IzzySoft** [comment]
Yeah, that caught me more than once as well. With Dropbox included, it would raise the `NonFreeNet` AntiFeature, though.

**Comment by IzzySoft** [comment]
Hey, cool that!

@yevhenii-kanivets Which way do you want to proceed? Shall I create a request for packaging with F-Droid – or do you rather want to keep the ball low and first provide both APKs and I take the "free" one into my repo, and we decide for the official one after that? Which way ever, we first need to wait for the next tag covering the PR.

@fynngodau thanks for your work! Would you like to create another PR establishing Fastlane structures for description and screenshots, if Yevhenii decides to go for the official repo? Details on that can be [found here](https://f-droid.org/en/docs/All_About_Descriptions_Graphics_and_Screenshots/#in-the-apps-source-repository), initial description/screenshots could be taken from Google Play, per-release changelogs are optional and could be skipped. Basically something like this:

```
└── fastlane
    └── metadata
        └── android
            ├── en-US
            │   ├── short_description.txt
            │   ├── full_description.txt
            │   ├── phoneScreenshots
            │   │   ├── 1.png
            │   │   ├── 2.png
            │   │   ...
            │   └── changelogs
            │       ├── 100000.txt
            │       └── 100100.txt
            └── ru
                ...
                └── changelogs
                    └── 100100.txt
```

@yevhenii-kanivets if you *install* fastlane (for F-Droid, the structures suffice), you could use that for Google Play as well – catching two birds with one stone: one place to maintain your app's metadata, distributed to both places. `en-US` would be mandatory for F-Droid (default language), other locales are optional.

A note on the screenshots: IMHO they ideally come without "framing", considering that people want to see and decipher them on small-screen devices. But the ones you use at Play would be OK, too.

**Comment by fynngodau** [comment]
If @yevhenii-kanivets does not want to use fastlane, it seems it is also possible to have fewer layers of directories (only a `metadata` folder, leaving out `fastlane` and `android` layers) [as fdroidclient does][fdroid]. I couldn't find any documentation on this structure, but the rest looks unchanged.

Furthermore I agree with @IzzySoft that unframed screenshots are preferable, but I can open a PR with the material currently available on Google Play anyway.

I think Money-Tracker still has no license, preventing it from being included in the main F-Droid repository (not sure what the policy for your repo is, @IzzySoft).

[fdroid]: https://gitlab.com/fdroid/fdroidclient/tree/master/metadata

**Comment by IzzySoft** [comment]
Doesn't hurt to keep it in fastlane structure and just not using the binaries. That way it's at least using a widely accepted standard. The other structure is rather F-Droid internal (you find it behind my link above, in the section describing to put it inside F-Droid metadata – which is what we meanwhile discourage, as it was growing to much and makes it hard to handle on our end. Ahem, time for disclosure: when I speak about "our", I'm one of the F-Droid maintainers :wink:).

**Comment by yevhenii-kanivets** [comment]
@IzzySoft @fynngodau sorry, I was busy this weekend (moving between countries). I'll think which option is better for OMT. For now, I think that fastlane is a better longterm solution. But let me think few days...

**Comment by IzzySoft** [comment]
Yes, Fastlane is definitely to be preferred: it's an established structure not only with F-Droid. And sure, whenever you're ready 😃 Just ping us when a release is tagged with the new build flavor, a license and potentially also Fastlane structures covered – and let us know where to go from there. As pointed out: I gladly help you getting your app into the official repo then – or, if you could provide the corresponding APK, take it in mine.

**Comment by IzzySoft** [comment]
Any news here, @yevhenii-kanivets?  I see you've made a new release yesterday. If you could also attach the `app-foss-release.apk`, I could add that to [my repo](https://apt.izzysoft.de/fdroid)¹ meanwhile – and as soon as you've decided on Fastlane¹ (and have it set-up as well as tagged), gladly help you to bring it into the official F-Droid repo.

¹ <sub>1+1=2: with that goal in mind, I could even prepare you the Fastlane data while establishing your app in my repo, and then upload them here for you to use. I've just done it that way for another app.</sub>

**Comment by yev-kanivets** [comment]
@IzzySoft I'm sorry for not keeping you informed. Yes, my plans is too fully released the new version of OMT to GP to the end of the week and then, while our developer is working on another project, configure fastlane. So should be done by the end of the month.

**Comment by IzzySoft** [comment]
No worries then – I was just wondering. My offer stands: As soon as you provide a FOSS APK, I can take that into my repo and prepare fastlane, provide that to you, and then you can "fine-tune" it. Of course if you prefer you can do it yourself :wink:

**Comment by yev-kanivets** [comment]
@IzzySoft thanks a lot 🙏It would be very nice of you if you can do fastlane. I'm not sure for what "FOSS" apk exactly stands for, but here I've uploaded both, free and proprietary builds - https://github.com/xorum-io/open_money_tracker/releases/tag/v2.1.0.

**Comment by IzzySoft** [comment]
FOSS stands for [Free and Open Source Software](https://en.wikipedia.org/wiki/Free_and_open-source_software) (sometimes also referred to as FLOSS, with the L standing for Libre – but that sounds too much like teeth cleaning stuff) – and in this context means "free of any proprietary components". And yes, I was referring to the APK you named `app-free-release.apk`, so thanks a lot!

I'll see that I set that up in my repo ASAP, going straight with Fastlane. Will then ZIP up the Fastlane structure and attach it here, so you can simply extract that into the root folder of your project and then simply do a `git add fastlane/* && git commit 'adding fastlane structures' && git push` if you are satisfied with the results. After having added a tag on top of that (and with my library scanner confirming all being fine), I then gladly assist you with the process of getting your app into the official repo.

**Comment by IzzySoft** [comment]
As promised, here comes the [fastlane.zip](https://github.com/yevhenii-kanivets/open_money_tracker/files/3643379/fastlane.zip). You can expand upon it (see [my snippet here](https://gitlab.com/snippets/1895688), where I put some basic notes on the fastlane structure, gathered from different places) – recommended would e.g. be per-release changelogs, and adding more languages is certainly welcome too :wink:

Note: I "de-framed" the screenshots. Reason: using a mobile client on small-screen devices, screen estate is little – and potential users should see as many details as possible; frames make that difficult without adding any useful information (users can just zoom-in to full screen, and have their own frame :wink:). I took the screenshots from your app's Playstore presence, so resolution is not the best; you probably want to replace them with something at least around 350x630px (a good compromise between "showing details" and "limiting transfer size / data volume").

My scanner confirms your app should be fit for the official repo (don't be confused about the `NonFreeNet`: that's because Dropbox is a proprietary *service* – but the Dropbox library itself is free AFAIK). All that's missing now is a license covering your app. What will it be?

Your app just [showed up here](https://apt.izzysoft.de/fdroid/index/apk/com.blogspot.e_kanivets.moneytracker), so you can see what the current Fastlane data (plus the F-Droid metadata file) turn into. Just let me know when you're ready (fastlane, license, tag) and I'll help your app into the official catalog then.

Thanks a lot again!

**Comment by yev-kanivets** [comment]
@IzzySoft thanks for your work 🙏 I'll try to setup fastlane, license and tag tomorrow evening.

**Comment by IzzySoft** [comment]
Cool!

And btw, more news from Mastodon – you got [MastoArt in Acryl](https://mastodon.technology/@mondstern/tagged/mastoart), [see here](https://mastodon.technology/@mondstern/102848449464393720) (Mondstern is our famous MastoArtist who awards selected projects with an Acryl drawing of their Identicon, to pay his respect to their authors/developers. You're welcome to use it (with credit to him)) :v:

**Comment by yev-kanivets** [comment]
@IzzySoft here is a [tag](https://github.com/xorum-io/open_money_tracker/releases/tag/v2.1.1) where I've included fastlane from your zip, license (MIT). Also I've added a link to your F-Droid repo to ReadMe page of OMT.

Thanks one more time for your help 🙏 

P.S. Agree about screenshots and translations, we will automate it through fastlane in future.

**Comment by IzzySoft** [comment]
Great, and thanks a lot! I've just [opened the RFP at the official repo](https://gitlab.com/fdroid/rfp/issues/1111). I hope our packagers there can deal with the globally declared Fabric stuff (removing it via `sed` or the likes). Let's see. Until solved there, it will be available via my repo :wink:

Should you update screenshots in Fastlane, please drop me a note: in my repo, Fastlane data isn't pulled automatically, so I need to check that by hand then. Same if you wish the description changed, or add new languages. Once in the main repo, this will be fully automated.

**Comment by IzzySoft** [comment]
@yevhenii-kanivets just wondering which repo F-Droid should build from and link to (for issues etc): this one here, or https://github.com/xorum-io/open_money_tracker – could you please specify?

**Comment by yev-kanivets** [comment]
@IzzySoft OK. Right repo is https://github.com/xorum-io/open_money_tracker.

**Comment by IzzySoft** [comment]
Good, thanks – then I put the correct one into the RFP. Just out of curiosity: why two repos at all?

**Comment by yev-kanivets** [comment]
@IzzySoft we (me and my partner) have started a company and taken OMT as a part of our portfolio. In return, we are supporting it and adding new features with much bigger speed than before :)
